core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java

```diff
@@ -20,42 +20,108 @@
+            "extended_bounds", a -> {
+        Long min = null;
+        Long max = null;
+        String minAsStr = null;
+        String maxAsStr = null;
+        if (a[0] == null) {
+            // nothing to do with it
+        } else if (a[0] instanceof Long) {
+            min = (Long) a[0];
+        } else if (a[0] instanceof String) {
+            minAsStr = (String) a[0];
+        } else {
+            throw new IllegalArgumentException("Unknown field type [" + a[0].getClass() + "]");
+        }
+        if (a[1] == null) {
+            // nothing to do with it
+        } else if (a[1] instanceof Long) {
+            max = (Long) a[1];
+        } else if (a[1] instanceof String) {
+            maxAsStr = (String) a[1];
+        } else {
+            throw new IllegalArgumentException("Unknown field type [" + a[1].getClass() + "]");
+        }
+        return new ExtendedBounds(min, max, minAsStr, maxAsStr);
+    });
+    static {
+        NoContextParser<Object> longOrString = p -> {
+            if (p.currentToken() == Token.VALUE_NUMBER) {
+                return p.longValue(false);
+            }
+            if (p.currentToken() == Token.VALUE_STRING) {
+                return p.text();
+            }
+            if (p.currentToken() == Token.VALUE_NULL) {
+                return null;
+            }
+            throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+        };
+        PARSER.declareField(optionalConstructorArg(), longOrString, MIN_FIELD, ValueType.LONG_OR_NULL);
+        PARSER.declareField(optionalConstructorArg(), longOrString, MAX_FIELD, ValueType.LONG_OR_NULL);
+    }
 
-    String minAsStr;
-    String maxAsStr;
+    /**
+     * Parsed min value. If this is null and {@linkplain #minAsStr} isn't then this must be parsed from {@linkplain #minAsStr}. If this is
+     * null and {@linkplain #minAsStr} is also null then there is no lower bound.
+     */
+    private final Long min;
+    /**
+     * Parsed min value. If this is null and {@linkplain #maxAsStr} isn't then this must be parsed from {@linkplain #maxAsStr}. If this is
+     * null and {@linkplain #maxAsStr} is also null then there is no lower bound.
+     */
+    private final Long max;
 
-    ExtendedBounds() {} //for parsing
+    private final String minAsStr;
+    private final String maxAsStr;
 
+    /**
+     * Construct with parsed bounds.
+     */
     public ExtendedBounds(Long min, Long max) {
-        this.min = min;
-        this.max = max;
+        this(min, max, null, null);
     }
 
+    /**
+     * Construct with unparsed bounds.
+     */
     public ExtendedBounds(String minAsStr, String maxAsStr) {
+        this(null, null, minAsStr, maxAsStr);
+    }
+
+    /**
+     * Construct with all possible information.
+     */
+    private ExtendedBounds(Long min, Long max, String minAsStr, String maxAsStr) {
+        this.min = min;
```

Comment:
maybe add assertion here for `(min == null) != (minAsStr == null);` and same for`max`.